doc/rfcs/2025-11-30--InterProcessCommunication.md

```diff
@@ -0,0 +1,440 @@
+stricter performance requirements.
+
+**One perfect mechanism.**
+We do not believe that there is any single best IPC mechanism that would
+support all application needs. Instead, we focus on an ecosystem of capsule
+implementations, with some basic mechanisms provided initially and others added
+as-needed. Additional capsules should be creatable downstream which could
+revise the interface of existing mechanisms, optimize existing mechanisms for a
+particular use case, or provide altogether new mechanisms.
+
+**Peer-to-Peer Discovery and Initiation.**
+The primary focus of the IPC design is to support client-server interactions
+where clients discover servers and initiate a connection. The current belief is
+that most applications will have a clear differentiation between servers which
+have resources and clients which take actions to connect with them. As an
+example of how this is applied to designs, in the current design servers
+"discover" clients when the client first contacts it, usually via the
+Synchronous Mailbox. An alternative design where any process could discover and
+reach out to any other process is possible, but is not the current focus. Note
+that this does not restrain communication patterns after initiation. Both
+client requests and asynchronous server notifications are supported in
+communication mechanisms.
+
+
+## IPC Name Registry Capsule
+
+This capsule provides registration for services and discovery for clients.
+Upon discovery, it provides opaque process IDs which can be used to refer to
+processes in other IPC mechanisms.
+
+If validation of services or clients is desired, this capsule could perform
+that operation at registration-time. The initial design will likely skip that
+feature but the design should enable it to be possible and provide clear
+locations where it could later be connected if implemented. Callbacks will be
+given on registration and discovery, allowing asynchronous operations to take
+place before they are completed.
+
+Registration and discovery is performed by matching "names", which are
+twenty-byte arrays. Allowed buffers must be exactly twenty bytes in length
+or commands using them will fail. The byte values can be anything, but are
+typically ASCII values, which do not need to be null-terminated. The default
+name value of all-zero-values cannot be used for discovery. Values will be
+copied from the allowed buffer into a fixed-size allocation in the grant
+region.
+
+An alternative mechanism of using package identifiers was considered, but was
+less flexible than having arbitrary process-specified identifiers. Package
+identifiers have some advantages that they are 1) arbitrary length, and 2) are
+unable to be modified at runtime. A downside of package identifiers is that
+they are encoded as part of the build system, and not part of the application
+code.
+
+**Commands**:
+* Existence
+* Register as service with allowed string name
+* Discover service with allowed string name
```

Comment:
How are races between service registration and service discovery handled (i.e. what happens if a process tries to discover service X when the process that provides service X is still starting up and hasn't registered yet)? Is that an implementation detail that is outside the scope of this document?

Reply:
We should document that in this document. Since servers can choose to register themselves, it is possible to search the registry and not discover a server yet. That would result in "no match". The application should try again at a later time.